### PR TITLE
feat: multi-tenant auth with per-user workspace isolation

### PR DESCRIPTION
**Pull request by standardtoaster**
**Rebased refile of #351** (closed in backlog triage). Previously reviewed by @serrrfirat and @zmanian — all review feedback was addressed. Rebased onto `staging` with no functional changes.

This addresses the same class of vulnerability as #760 (thread_id context pollution) architecturally — when every request is scoped to an authenticated user_id via `GATEWAY_USER_TOKENS`, cross-user pollution can't occur regardless of the attack vector.

This PR includes 39 HTTP-level integration tests for auth, isolation, and ownership checks — including DB-backed job ownership tests using in-memory libSQL. These don't map naturally to the trajectory format (auth happens before the agent loop), but happy to discuss what multi-tenant trajectory coverage should look like.

Broader context — I'm building a multi-user personal AI assistant on IronClaw and want to make sure I'm contributing in a useful direction. Would be great to sync on priorities if there's a good channel for that.

Depends on #1117.

Original PR: #351

---

> **Part 3 of 3** for Issue #59 (multi-tenancy). Depends on #1112 and #1117 — merge those first. The diff here includes all three PRs; once the first two merge, the diff shrinks to ~1,300 lines across 21 files.

## Summary

Adds token-based multi-user authentication to the web gateway, giving each
user a fully isolated workspace with independent memory layers and
cross-scope read access. Builds on the layered memory (#1112) and
multi-scope reads (#1117) to deliver end-to-end multi-tenant workspace
isolation.

### How it works

Single-user mode is the default and behaves identically to today. Multi-user
mode activates only when `GATEWAY_USER_TOKENS` is set:

```
GATEWAY_USER_TOKENS='{"tok-alice":{"user_id":"alice","workspace_read_scopes":["shared"]},"tok-bob":{"user_id":"bob","workspace_read_scopes":["shared"]}}'
```

Each user gets:
- **Isolated workspace writes** — scoped by `user_id` at the DB level
- **Independent memory layers** with privacy redirect (from #1112)
- **Optional cross-scope reads** (from #1117)
- **Scoped SSE/WebSocket streams** — only see your own events
- **Independent rate limiting** — one user can't starve another

### Key components

| Component | Purpose |
|-----------|---------|
| `MultiAuthState` | Maps bearer tokens → `UserIdentity` (user_id, read scopes, memory layers) |
| `AuthenticatedUser` | Axum extractor that provides the resolved identity to handlers |
| `WorkspacePool` | Lazily creates and caches per-user workspaces with double-checked locking |
| `PerUserRateLimiter` | Independent sliding-window rate limits per user_id |
| `ScopedEvent` | SSE envelope with optional `user_id`; subscribers filter to their own events |

## Security hardening

After the initial implementation, three rounds of AI-assisted code review
identified shared resources that become cross-tenant data leaks once
multiple users share a gateway. These were pre-existing on upstream but
harmless in single-user mode — multi-tenancy is what makes them
exploitable.

### Fixed in this PR

| Issue | Fix |
|-------|-----|
| SSE broadcast to all subscribers | `ScopedEvent` envelope; subscribers filter by user_id |
| Single shared rate limiter | `PerUserRateLimiter` with independent windows per user |
| Routine handlers had no auth | `AuthenticatedUser` + `routine.user_id` ownership check |
| Job prompt handler skipped ownership when store=None | Require store (503) |
| SSE/WS subscribe was unscoped | Pass authenticated user_id to subscribe filter |
| OpenAI compat used `default_user_id` for rate limiting | Extract `AuthenticatedUser`, use `user.user_id` |
| IPv6 WebSocket origin validation | Extract `is_local_origin()` helper with bracket handling |
| `conversation_belongs_to_user` returned false on DB error | Propagate error as 500 instead of masking |
| `sandbox_job_belongs_to_user` returned false on DB error | Same (2 instances) |
| `PerUserRateLimiter` panicked on lock poisoning | `into_inner()` recovery |
| WS approval delivery was fire-and-forget | Send error to client on failure |
| `jobs_detail_handler` swallowed DB errors as 404 | Propagate as 500 |
| `jobs_cancel_handler` swallowed DB errors as 404 | Same |
| `get_sandbox_job_mode` silently defaulted to Worker on DB error | Log warning, then default |
| `chat_threads_handler` silently dropped DB errors | Log error before in-memory fallback |
| `send_status` silently broadcast globally when user_id missing | Added debug log |
| `UserTokenConfig` empty user_id | Validation in config parsing |

### Known limitations (documented, need broader changes)

| Issue | Required change |
|-------|-----------------|
| Sandbox job SSE events broadcast to all tenants | Orchestrator needs per-job user_id tracking |
| Process-wide log stream shared across tenants | Needs per-user filtering or RBAC |
| Extension auth/status broadcasts are global | Extensions need user context threaded through |

## Changes (this PR only)

| File | What |
|------|------|
| `auth.rs` | `MultiAuthState`, `UserIdentity`, `AuthenticatedUser` extractor, case-insensitive Bearer parsing |
| `server.rs` | `WorkspacePool`, `PerUserRateLimiter`, `resolve_workspace()`, handler auth/ownership, `is_local_origin()` |
| `sse.rs` | `ScopedEvent` envelope, `broadcast_for_user()`, user-scoped `subscribe()`/`subscribe_raw()` |
| `ws.rs` | Pass user_id to subscribe, scope auth broadcasts, approval error reporting |
| `mod.rs` | `new_multi_auth()`, `with_workspace_pool()`, scoped channel broadcasts |
| `openai_compat.rs` | Per-user rate limiter extraction |
| `config/channels.rs` | `GATEWAY_USER_TOKENS` parsing, `UserTokenConfig`, validation |
| `extensions/manager.rs` | Document user-scoping limitations |
| `main.rs` | Multi-user auth state + workspace pool wiring |
| `test_helpers.rs` | `TestGatewayBuilder::start_multi()` for multi-user server tests |
| `tests/multi_tenant_integration.rs` | 39 integration tests (see below) |
| `tests/openai_compat_integration.rs` | Updated for new `GatewayState` fields |
| `tests/ws_gateway_integration.rs` | Updated for new `GatewayState` fields |
| `tests/support/gateway_workflow_harness.rs` | Updated for new `GatewayState` fields |

## Integration test coverage (39 tests)

**Unit-level (22):** MultiAuthState token→identity mapping, query token auth restrictions, per-user rate limiting isolation, SSE event scoping (user A can't see user B's events), edge cases (empty token, prefix match, first_token/first_identity).

**Handler-level without DB (12):** Full HTTP stack through real Axum server — protected endpoints reject unauthenticated/unknown tokens, public health endpoint accessible without auth, chat send flows through auth to agent channel, query token accepted on SSE but rejected on non-SSE endpoints, WebSocket per-user event isolation (Alice's scoped events not visible to Bob's WS connection).

**Handler-level with DB (5):** In-memory libSQL — Alice lists jobs and sees only hers, Bob requests Alice's job by ID and gets 404 (not 403, preventing enumeration), Alice can see her own job detail, Bob lists jobs and sees only his, nonexistent job returns 404.

## Test plan

- [x] `cargo clippy --all --benches --tests --examples` — zero warnings
- [x] `cargo test --lib` — 3025 tests pass
- [x] `cargo test --test multi_tenant_integration` — 39 tests pass
- [x] Single-user mode unaffected (no `GATEWAY_USER_TOKENS` → identical behavior)

### CODE REVIEW

File: src/channels/web/server.rs [hunk 1]
```diff
@@ -148,14 +153,116 @@ pub struct ActiveConfigSnapshot {
...
+        }
+
+        // Slow path: create workspace under write lock
+        let mut cache = self.cache.write().await;
+        // Double-check after acquiring write lock
+        if let Some(ws) = cache.get(&identity.user_id) {
+            return Arc::clone(ws);
+        }
+
+        let mut ws = Workspace::new_with_db(&identity.user_id, Arc::clone(&self.db));
```
Comment (serrrfirat) [inline src/channels/web/server.rs]: **High Severity** – Token read scopes are ignored

`WorkspacePool::get_or_create` never applies `identity.workspace_read_scopes` when building the per-user workspace. Every workspace in the cache is created with just `[user_id]`, so even if `GATEWAY_USER_TOKENS` grants a token access to `"shared"`, calls like `resolve_workspace()` return a workspace that cannot see that scope – cross-scope reads advertised in the PR never work. Consider calling `ws = ws.with_additional_read_scopes(identity.workspace_read_scopes.clone())` (with dedup) before caching, and add an integration test that proves Alice can read from the shared scope when her token config includes it.

File: src/channels/web/server.rs [hunk 2]
```diff
@@ -148,14 +153,116 @@ pub struct ActiveConfigSnapshot {
...
+            .or_insert_with(|| RateLimiter::new(self.max_requests, self.window_secs));
+        limiter.check()
+    }
+}
+
+/// Per-user workspace pool: lazily creates and caches workspaces keyed by user_id.
+///
+/// In single-user mode, exactly one workspace is cached. In multi-user mode,
+/// each authenticated user gets their own workspace with appropriate scopes.
+pub struct WorkspacePool {
```
Comment (serrrfirat) [inline src/channels/web/server.rs:220]: **Medium Severity** – Multi-tenant workspaces drop the configured layers/search settings

`AppBuilder` applies `with_search_config`, `with_additional_read_scopes` (from `WorkspaceConfig`), memory layers, and any privacy classifier when it builds the single-user workspace (see `src/app.rs:326-343`). When the gateway is run in multi-user mode, `resolve_workspace()` goes through this pool instead, but `WorkspacePool::get_or_create` only does `Workspace::new_with_db` (+ embeddings) and caches that. The per-user workspaces therefore lose all operator-provided configuration (search tuning, global read scopes, layer definitions, privacy classifier) as soon as `workspace_pool` is used. Could we thread the same config into the pool (e.g., pass `WorkspaceConfig`, search settings, classifier into `WorkspacePool::new` and apply them before caching), so tenants behave the same regardless of single- vs multi-user mode? A regression test asserting that multi-tenant workspaces inherit the configured read scopes would make this concrete.

File: src/channels/web/handlers/jobs.rs [hunk 3]
```diff
@@ -80,6 +85,7 @@ pub async fn jobs_list_handler(
 
 pub async fn jobs_summary_handler(
     State(state): State<Arc<GatewayState>>,
+    AuthenticatedUser(_user): AuthenticatedUser,
```
Comment (serrrfirat) [inline src/channels/web/handlers/jobs.rs]: **High Severity** Multi-tenant leak in `/api/jobs/summary`

`jobs_summary_handler` now receives the authenticated user, but it still calls `sandbox_job_summary()` / `agent_job_summary()` which each return global counters. In multi-tenant mode that means Alice can see how many sandbox/agent jobs Bob has queued just by hitting `/api/jobs/summary`. Please scope both queries by `user.user_id` (there is already `sandbox_job_summary_for_user`, and the agent counts can be derived from `list_agent_jobs_for_user` or a new summary query) so the endpoint only reports the caller's own jobs.

File: src/channels/web/handlers/jobs.rs [hunk 4]
```diff
@@ -476,13 +531,34 @@ pub async fn jobs_restart_handler(
...
     Path(id): Path<String>,
     Json(body): Json<serde_json::Value>,
 ) -> Result<Json<serde_json::Value>, (StatusCode, String)> {
     let job_id: uuid::Uuid = id
         .parse()
         .map_err(|_| (StatusCode::BAD_REQUEST, "Invalid job ID".to_string()))?;
 
+    // Verify ownership before queuing a prompt.
+    if let Some(ref store) = state.store {
+        match store.get_sandbox_job(job_id).await {
```
Comment (serrrfirat) [inline src/channels/web/handlers/jobs.rs]: **High Severity** Agent prompts always return 404**

The new sandbox ownership check in `jobs_prompt_handler` returns `404` whenever `get_sandbox_job(job_id)` yields `None`. That happens for every agent job, so the function now bails out before the scheduler code path runs and it is impossible for any user to send follow-up prompts to running agent jobs. Please only enforce the sandbox ownership check when a sandbox record exists, and then look up the agent job (verifying `ctx.user_id == user.user_id`) before calling the scheduler path.

File: src/channels/web/handlers/jobs.rs [hunk 5]
```diff
@@ -145,176 +152,221 @@ pub async fn jobs_detail_handler(
...
-            "job_id": job_id,
-        })));
     }
 
     Err((StatusCode::NOT_FOUND, "Job not found".to_string()))
 }
 
 pub async fn jobs_restart_handler(
     State(state): State<Arc<GatewayState>>,
+    AuthenticatedUser(user): AuthenticatedUser,
```
Comment (serrrfirat) [inline src/channels/web/handlers/jobs.rs:366]: **High Severity** Agent restart missing ownership check**

Thanks for threading `AuthenticatedUser` through `jobs_restart_handler`, but the agent branch (`store.get_job`) still dispatches a restart without verifying `old_job.user_id == user.user_id`. Any tenant who guesses a UUID can now restart someone else's agent job (and re-run their workload) through `/api/jobs/{id}/restart`. Please mirror the sandbox branch: fetch the agent job, compare the owner, and return 404 when it doesn't belong to the caller.
